### PR TITLE
Upgrade to 5.3 - Configuration Changes

### PR DESCRIPTION
**Pull request by mmoayyed**

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -37,10 +53,11 @@ javaxServletVersion=3.1.0
 javaxElVersion=3.0.0
 javaxElImplVersion=2.2.6
 javaxValidationVersion=1.1.0.Final
-javassistVersion=3.21.0-GA
+javassistVersion=3.22.0-GA
 javaxJstlVersion=1.2
 
-semverVersion=2.0.3
+semverVersion=2.1.0
```
Comment (mmoayyed) [inline gradle.properties]: I am not quite if all of these versions are in fact needed, and/or the dependencies. Very likely, you should try to remove as many of them as possible.

File: gradle/dependencies.gradle [hunk 2]
```diff
@@ -484,9 +502,14 @@ ext.libraries = [
                     force = true
                 }
         ],
-        hazelcast               : dependencies.create("com.hazelcast:hazelcast:$hazelcastVersion") {
-            force = true
-        },
+        hazelcast               : [
```
Comment (mmoayyed) [inline gradle/dependencies.gradle]: See above. All such dependencies really need to go away.

File: gradle/wrapper/gradle-wrapper.properties [hunk 3]
```diff
@@ -1,5 +1,6 @@
+#Thu Dec 07 09:02:34 PST 2017
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.3.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.3.1-all.zip
```
Comment (mmoayyed) [inline gradle/wrapper/gradle-wrapper.properties]: Probably switch to 4.4

File: webapp-mgmt/cas-management-webapp-configuration/build.gradle [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    implementation "org.apereo.cas:cas-server-core-api-configuration:$casVersion"
+    implementation "org.apereo.cas:cas-server-core-util:$casVersion"
+    implementation "org.apereo.cas:cas-server-core-configuration-api:$casVersion"
+
+}
+
+task generateConfigurationMetadata(type: JavaExec) {
+    doFirst {
+        println "\nRunning CAS Spring configuration metadata processor in addition to the native " +
```
Comment (mmoayyed) [inline webapp-mgmt/cas-management-webapp-configuration/build.gradle]: I'll take a look at this bit later. Fine as is. @hdeadman indicated there may be an issue with the generator not honoring parent classes. If I get that fixed, I'll run over to the mgmt app and see if I can get it to also generate metadata correctly.

File: webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasConfigurationPropertiesEnvironmentManager.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+import java.io.File;
+import java.util.Map;
+
+/**
+ * This is {@link CasConfigurationPropertiesEnvironmentManager}.
+ *
+ * @author Misagh Moayyed
+ * @since 5.1.0
+ */
+public class CasConfigurationPropertiesEnvironmentManager {
```
Comment (mmoayyed) [inline webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasConfigurationPropertiesEnvironmentManager.java]: This file need to go away. You should use one that is provided to you by the configuration-api module.

File: webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasConfigurationPropertiesSourceLocator.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
+package org.apereo.cas.mgmt.configuration;
+
+import org.apache.commons.io.FileUtils;
```
Comment (mmoayyed) [inline webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasConfigurationPropertiesSourceLocator.java]: This file need to go away. You should use one that is provided to you by the configuration-api module.

File: webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasCoreBootstrapStandaloneConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.core.env.Environment;
+import org.springframework.core.env.PropertySource;
+import org.springframework.core.io.ResourceLoader;
+
+import java.util.List;
+
+/**
+ * This is {@link CasCoreBootstrapStandaloneConfiguration}.
+ *
+ * @author Misagh Moayyed
```
Comment (mmoayyed) [inline webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasCoreBootstrapStandaloneConfiguration.java]: Author tags likely need to be revisited.

File: webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasManagementConfigurationProperties.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.context.properties.NestedConfigurationProperty;
+
+@ConfigurationProperties("cas")
+public class CasManagementConfigurationProperties {
+    /**
+       Authentication
+     */
+    @NestedConfigurationProperty
+    AuthenticationProperties authn = new AuthenticationProperties();
```
Comment (mmoayyed) [inline webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasManagementConfigurationProperties.java]: Are these settings/properties used anywhere here in the management app?

File: webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CommaSeparatedStringToThrowablesConverter.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This is {@link CommaSeparatedStringToThrowablesConverter}.
+ *
+ * @author Misagh Moayyed
+ * @since 5.3.0
+ */
+public class CommaSeparatedStringToThrowablesConverter implements Converter<String, List<Class<? extends Throwable>>> {
```
Comment (mmoayyed) [inline webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CommaSeparatedStringToThrowablesConverter.java]: This file need to go away. You should use one that is provided to you by the configuration-api module.

**Comment by mmoayyed** [comment]
Generally, looks fine. There is some code duplication in the configuration area that needs to be looked at and removed.

File: webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasConfigurationPropertiesEnvironmentManager.java (hunk 5 above)
Comment (tsschmidt) [reply webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasConfigurationPropertiesEnvironmentManager.java]: What about the rebind method that uses CasConfigurationProperties?

File: webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasManagementConfigurationProperties.java (hunk 8 above)
Comment (tsschmidt) [reply webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasManagementConfigurationProperties.java]: Yes, I believe mostly by the ldap module

File: webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasConfigurationPropertiesEnvironmentManager.java (hunk 5 above)
Comment (mmoayyed) [reply webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasConfigurationPropertiesEnvironmentManager.java]: That I don't think you need either...or at least not right away. rebind is mostly used to pick up changes done to the config file and dynamically rebind objects to values and refresh the application context. I'd say that's not a very immediate use case for the management app.

File: webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasManagementConfigurationProperties.java (hunk 8 above)
Comment (mmoayyed) [reply webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasManagementConfigurationProperties.java]: Hmm...confused here! Why would the management app want to deal with anything authentication-related? Are you referring to ways one get to the management app? I dont recall doing anything ever in the management.properties file that would take the form of `cas.authn.something`. Am I missing something here?

Comment (tsschmidt) [reply webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasManagementConfigurationProperties.java]: It is used here:
``` @RefreshScope
    @ConditionalOnMissingBean(name = "attributeRepository")
    @Bean
    public IPersonAttributeDao attributeRepository() {
        return Beans.newStubAttributeRepository(casProperties.getAuthn().getAttributeRepository());
    }
```
and Here
```
    @RefreshScope
    @Bean
    public Collection<BaseOidcScopeAttributeReleasePolicy> userDefinedScopeBasedAttributeReleasePolicies() {
        final OidcProperties oidc = casProperties.getAuthn().getOidc();
        return oidc.getUserDefinedScopes().entrySet()
                .stream()
                .map(k -> new OidcCustomScopeAttributeReleasePolicy(k.getKey(), CollectionUtils.wrapList(k.getValue().split(","))))
                .collect(Collectors.toSet());
    }
```

Comment (mmoayyed) [reply webapp-mgmt/cas-management-webapp-configuration/src/main/java/org/apereo/cas/mgmt/configuration/CasManagementConfigurationProperties.java]: Ah, gotcha. 

So, mostly suggestions then:

1. So the stub attribute repository, that just needs a map right? I wonder if we can remove the reference in the ctor and simply introduce a new setting in the `ManagementProperties` file. Don't know. This feels like the wrong place to define attributes used in the UI only, under an "authn" category. 

2. Same thing with the OIDC user scoped. All we need is a key and value, right? How about we just define a map directly and forego the `OidcProperties`.

This is perfectly good as is; I am mostly nit-picking really.

**Comment by tsschmidt** [comment]
Closed in favor of PR #4

**Pull request closed without merging**